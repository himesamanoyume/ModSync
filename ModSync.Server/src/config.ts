import path from "node:path";
import type { PreSptModLoader } from "@spt/loaders/PreSptModLoader";
import type { JsonUtil } from "@spt/utils/JsonUtil";
import type { VFS } from "@spt/utils/VFS";
import { glob } from "./utility/glob";
import { unixPath } from "./utility/misc";
import type { ILogger } from "@spt/models/spt/utils/ILogger";

export type SyncPath = {
	path: string;
	enabled?: boolean;
	enforced?: boolean;
	silent?: boolean;
	restartRequired?: boolean;
};

type RawConfig = {
	syncPaths: (string | SyncPath)[];
	exclusions: string[];
};

const DEFAULT_CONFIG = `{
	"syncPaths": [
		"BepInEx/plugins",
		"BepInEx/patchers",
		"BepInEx/config",
		{
			"enabled": false,
			"path": "user/mods",
			"restartRequired": false
		}
	],
	"exclusions": [
		// SPT Installer
		"BepInEx/plugins/spt",
		"BepInEx/patchers/spt-prepatch.dll",
		// Questing Bots
		"BepInEx/plugins/DanW-SPTQuestingBots/log",
		// Realism
		"user/mods/SPT-Realism/ProfileBackups",
		// Fika
		"user/mods/fika-server/types",
		"user/mods/fika-server/cache",
		"BepInEx/plugins/Fika.Dedicated.dll",
		// Live Flea Prices
		"user/mods/zzDrakiaXYZ-LiveFleaPrices/config",
		// Questing Bots
		"BepInEx/plugins/DanW-SPTQuestingBots/log",
		// EFTApi
		"BepInEx/plugins/kmyuhkyuk-EFTApi/cache",
		// Expanded Task Text (Accounts for bug with current version)
		"user/mods/ExpandedTaskText/src/**/cache.json",
		// Leaves Loot Fuckery
		"user/mods/leaves-loot_fuckery/output",
		// ADD MISSING QUEST WEAPON REQUIREMENTS
		"user/mods/zz_guiltyman-addmissingquestweaponrequirements/log.log",
		"user/mods/zz_guiltyman-addmissingquestweaponrequirements/user/logs",
		// Corter ModSync
		"BepInEx/patchers/Corter-ModSync-Patcher.dll",
		"**/*.nosync",
		"**/*.nosync.txt",
		// General server mods
		"user/mods/**/.git",
		"user/mods/**/node_modules",
		"user/mods/**/*.js",
		"user/mods/**/*.js.map",
		"**/*:Zone.Identifier"
	]
}`;

export class Config {
	private _globs: RegExp[];
	constructor(
		public syncPaths: Required<SyncPath>[],
		public exclusions: string[],
	) {
		this._globs = exclusions.map(glob);
	}

	public isExcluded(filePath: string): boolean {
		return this._globs.some(
			(glob) =>
				glob.test(unixPath(filePath)),
		);
	}
}
export class ConfigUtil {
	constructor(
		private vfs: VFS,
		private jsonUtil: JsonUtil,
		private modImporter: PreSptModLoader,
		private logger: ILogger,
	) { }

	/**
	 * @throws {Error} If the config file does not exist
	 */
	private async readConfigFile(): Promise<RawConfig> {
		const modPath = this.modImporter.getModPath("Corter-ModSync");
		const configPath = path.join(modPath, "config.jsonc");

		if (!this.vfs.exists(configPath))
			await this.vfs.writeFilePromisify(configPath, DEFAULT_CONFIG);

		return this.jsonUtil.deserializeJsonC(
			// @ts-expect-error I am right, SPT is wrong
			await this.vfs.readFilePromisify(configPath, { encoding: "utf-8" }),
			"config.jsonc",
		) as RawConfig;
	}

	/**
	 * @throws {Error} If the config is invalid
	 */
	private validateConfig(config: RawConfig): void {
		if (!Array.isArray(config.syncPaths))
			throw new Error(
				"Corter-ModSync: config.jsonc 'syncPaths' is not an array. Please verify your config is correct and try again.",
			);

		if (!Array.isArray(config.exclusions))
			throw new Error(
				"Corter-ModSync: config.jsonc 'exclusions' is not an array. Please verify your config is correct and try again.",
			);


		const uniquePaths = new Set();
		for (const syncPath of config.syncPaths) {
			if (typeof syncPath !== "string" && !("path" in syncPath))
				throw new Error(
					"Corter-ModSync: config.jsonc 'syncPaths' is missing 'path'. Please verify your config is correct and try again.",
				);

			if (
				typeof syncPath === "string"
					? path.isAbsolute(syncPath)
					: path.isAbsolute(syncPath.path)
			)
				throw new Error(
					`Corter-ModSync: SyncPaths must be relative to SPT server root. Invalid path '${syncPath}'`,
				);

			if (
				path
					.relative(
						process.cwd(),
						path.resolve(
							process.cwd(),
							typeof syncPath === "string" ? syncPath : syncPath.path,
						),
					)
					.startsWith("..")
			)
				throw new Error(
					`Corter-ModSync: SyncPaths must within SPT server root. Invalid path '${syncPath}'`,
				);

			if (uniquePaths.has(typeof syncPath === "string" ? syncPath : syncPath.path))
				throw new Error(
					`Corter-ModSync: SyncPaths must be unique. Duplicate path '${syncPath}'`,
				)
		}
	}

	public async load(): Promise<Config> {
		const rawConfig = await this.readConfigFile();
		this.validateConfig(rawConfig);

		return new Config(
			[
				{
					enabled: true,
					enforced: true,
					silent: true,
					restartRequired: false,
					path: "ModSync.Updater.exe",
				},
				{
					enabled: true,
					enforced: true,
					silent: true,
					restartRequired: true,
					path: "BepInEx/plugins/Corter-ModSync.dll",
				},
				...rawConfig.syncPaths
					.map((syncPath) => ({
						enabled: true,
						enforced: false,
						silent: false,
						restartRequired: true,
						...(typeof syncPath === "string" ? { path: syncPath } : syncPath),
					}))
					.sort((a, b) => b.path.length - a.path.length),
			],
			rawConfig.exclusions,
		);
	}
}
