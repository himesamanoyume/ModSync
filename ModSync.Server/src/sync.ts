import path from "node:path";
import type { ILogger } from "@spt/models/spt/utils/ILogger";
import type { VFS } from "@spt/utils/VFS";
import type { Config, SyncPath } from "./config";
import { hashFile } from "./utility/imoHash";
import { HttpError, winPath } from "./utility/misc";
import { Semaphore } from "./utility/semaphore";

type ModFile = {
	hash: string;
	directory: boolean;
};

export class SyncUtil {
	private limiter = new Semaphore(1024);

	constructor(
		private vfs: VFS,
		private config: Config,
		private logger: ILogger,
	) {}

	private async getFilesInDir(baseDir: string, dir: string): Promise<string[]> {
		if (!this.vfs.exists(dir)) {
			this.logger.warning(
				`Corter-ModSync: Directory '${dir}' does not exist, will be ignored.`,
			);
			return [];
		}

		const stats = await this.vfs.statPromisify(dir);
		if (stats.isFile()) return [dir];

		const files: string[] = [];
		for (const fileName of this.vfs.getFiles(dir)) {
			const file = path.join(dir, fileName);

			if (this.config.isExcluded(file)) continue;

			files.push(file);
		}

		for (const dirName of this.vfs.getDirs(dir)) {
			const subDir = path.join(dir, dirName);

			if (this.config.isExcluded(subDir)) continue;

			const subFiles = await this.getFilesInDir(baseDir, subDir);
			if (
				this.vfs.getFiles(subDir).length === 0 &&
				this.vfs.getDirs(subDir).length === 0
			)
				files.push(subDir);

			files.push(...subFiles);
		}

		if (
			stats.isDirectory() &&
			this.vfs.getFiles(dir).length === 0 &&
			this.vfs.getDirs(dir).length === 0
		)
			files.push(dir);

		return files;
	}

	private async buildModFile(
		file: string,
		// biome-ignore lint/correctness/noEmptyPattern: <explanation>
		{}: Required<SyncPath>,
	): Promise<ModFile> {
		const stats = await this.vfs.statPromisify(file);
		if (stats.isDirectory()) return { hash: "", directory: true };

		let retryCount = 0;
		const lock = await this.limiter.acquire();
		while (true) {
			try {
				const hash = await hashFile(file);
				lock.release();

				return {
					hash,
					directory: false,
				};
			} catch (e) {
				if (
					e instanceof Error &&
					"code" in e &&
					e.code === "EBUSY" &&
					retryCount < 5
				) {
					this.logger.error(
						`Error reading '${file}'. Retrying (${retryCount}/5)...`,
					);
					await new Promise((resolve) => setTimeout(resolve, 500));
					retryCount++;
					continue;
				}

				this.logger.error(`Error reading '${file}'. Exiting...`);
				this.logger.error(`${e}`);
				throw new HttpError(
					500,
					`Corter-ModSync: Error reading '${file}'\n${e}`,
				);
			}
		}
	}

	public async hashModFiles(
		syncPaths: Config["syncPaths"],
	): Promise<Record<string, Record<string, ModFile>>> {
		const result: Record<string, Record<string, ModFile>> = {};
		const processedFiles = new Set<string>();

		const startTime = performance.now();
		for (const syncPath of syncPaths) {
			const files = await this.getFilesInDir(syncPath.path, syncPath.path);
			const filesResult: Record<string, ModFile> = {};

			for (const file of files) {
				if (processedFiles.has(winPath(file))) continue;

				filesResult[winPath(file)] = await this.buildModFile(file, syncPath);

				processedFiles.add(winPath(file));
			}

			result[winPath(syncPath.path)] = filesResult;
		}

		this.logger.info(
			`Corter-ModSync: Hashed ${processedFiles.size} files in ${performance.now() - startTime}ms`,
		);

		return result;
	}

	/**
	 * @throws {Error} If file path is invalid
	 */
	public sanitizeDownloadPath(
		file: string,
		syncPaths: Config["syncPaths"],
	): string {
		const normalized = path.join(
			path.normalize(file).replace(/^(\.\.(\/|\\|$))+/, ""),
		);

		for (const syncPath of syncPaths) {
			const fullPath = path.join(process.cwd(), syncPath.path);
			if (!path.relative(fullPath, normalized).startsWith("..")) {
				return normalized;
			}
		}

		throw new HttpError(
			400,
			`Corter-ModSync: Requested file '${file}' is not in an enabled sync path!`,
		);
	}
}
