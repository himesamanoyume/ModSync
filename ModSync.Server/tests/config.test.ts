import { expect, beforeEach, describe, it, vi } from "vitest";

import { Config, ConfigUtil } from "../src/config";
import { VFS } from "./utils/vfs";
import { JsonUtil } from "./utils/jsonUtil";
import { PreSptModLoader } from "./utils/preSptModLoader";
import type { VFS as IVFS } from "@spt/utils/VFS";
import type { JsonUtil as IJsonUtil } from "@spt/utils/JsonUtil";
import type { PreSptModLoader as IPreSptModLoader } from "@spt/loaders/PreSptModLoader";
import { fs, vol } from "memfs";
import type { ILogger } from "@spt/models/spt/utils/ILogger";
import { mock } from "vitest-mock-extended";
import path from "node:path";

vi.mock("node:fs", async () => {
	const path = await vi.importActual("node:path");
	const { readFileSync } = await vi.importActual("node:fs");
	const { fs } = await vi.importActual("memfs");

	// @ts-expect-error Vitest types aren't exactly perfect
	const wasmPath = path.resolve(__dirname, "../src/utility/metrohash.wasm");

	// @ts-expect-error Vitest types aren't exactly perfect
	fs.writeFileSync(wasmPath, readFileSync(wasmPath))

	return fs;
});

describe("Config", () => {
	let config: Config;
	beforeEach(() => {
		config = new Config(
			[
				{
					path: "plugins",
					enabled: true,
					enforced: false,
					silent: false,
					restartRequired: true,
				},
				{
					path: "mods",
					enabled: false,
					enforced: false,
					silent: false,
					restartRequired: false,
				},
			],
			["plugins/**/node_modules", "plugins/**/*.js"],
		);
	});

	it("should correctly identify excluded paths", () => {

		expect(config.isExcluded("plugins/test.dll")).toBe(false);
		expect(config.isExcluded("plugins/banana/node_modules")).toBe(true);
		expect(config.isExcluded("plugins/banana/test.js")).toBe(true);
		expect(config.isExcluded("plugins/banana/config.json")).toBe(false);
		expect(config.isExcluded("plugins/banana/node_modules/lodash")).toBe(false);
	});
});

describe("ConfigUtil", () => {
	beforeEach(() => {
		vol.reset();
	});

	it("should create config if none exists", async () => {
		vol.fromNestedJSON(
			{ "placeholder.txt": "to ensure directory exists" },
			process.cwd(),
		);

		await new ConfigUtil(
			new VFS() as IVFS,
			new JsonUtil() as IJsonUtil,
			new PreSptModLoader() as IPreSptModLoader,
			mock<ILogger>(),
		).load();

		expect(fs.existsSync(path.join(process.cwd(), "config.jsonc"))).toBe(true);
	});

	it("should load config", async () => {
		vol.fromNestedJSON(
			{
				"config.jsonc": `{
				"syncPaths": [
					"plugins",
					{ "path": "mods", "enabled": false },
					{ "path": "doesnotexist", "enabled": true }
				],
				// Exclusions for commonly used SPT mods
				"exclusions": [
					"plugins/**/node_modules"
				]
			}`,
				plugins: {},
				mods: {},
			},
			process.cwd(),
		);

		const config = await new ConfigUtil(
			new VFS() as IVFS,
			new JsonUtil() as IJsonUtil,
			new PreSptModLoader() as IPreSptModLoader,
			mock<ILogger>(),
		).load();

		expect(config.syncPaths).toMatchSnapshot();

		expect(config.exclusions).toEqual(["plugins/**/node_modules"]);
	});

	it("Should always include ModSync.Updater.exe in syncPaths", async () => {
		vol.fromNestedJSON(
			{
				"config.jsonc": `{
					"syncPaths": [],
					// Exclusions for commonly used SPT mods
					"exclusions": [
						"plugins/**/node_modules"
					]
				}`,
				plugins: {},
				mods: {},
			},
			process.cwd(),
		);

		const config = await new ConfigUtil(
			new VFS() as IVFS,
			new JsonUtil() as IJsonUtil,
			new PreSptModLoader() as IPreSptModLoader,
			mock<ILogger>(),
		).load();

		expect(config.syncPaths).toMatchSnapshot();
	});

	it("should reject syncPath object without path", () => {
		vol.fromNestedJSON(
			{
				"config.jsonc": `{
					"syncPaths": [
						{ "enabled": true, "enforced": false, "path": "plugins", "restartRequired": true, "silent": false },
						{ "enabled": false, "enforced": false, "path": "mods", "restartRequired": true, "silent": false },
						{ "enabled": true, "enforced": false, "restartRequired": true, "silent": false }
					],
					// Exclusions for commonly used SPT mods
					"exclusions": [
						"plugins/**/node_modules"
					]
				}`,
			},
			process.cwd(),
		);

		const configUtil = new ConfigUtil(
			new VFS() as IVFS,
			new JsonUtil() as IJsonUtil,
			new PreSptModLoader() as IPreSptModLoader,
			mock<ILogger>(),
		);

		expect(configUtil.load()).rejects.toThrowErrorMatchingSnapshot();
	});

	it("should reject absolute paths", () => {
		vol.fromNestedJSON(
			{
				"config.jsonc": `{
					"syncPaths": [
						"/etc/shadow",
						"C:\\\\Windows\\\\System32\\\\cmd.exe"
					],
					// Exclusions for commonly used SPT mods
					"exclusions": [
						"plugins/**/node_modules"
					]
				}`,
			},
			process.cwd(),
		);

		const configUtil = new ConfigUtil(
			new VFS() as IVFS,
			new JsonUtil() as IJsonUtil,
			new PreSptModLoader() as IPreSptModLoader,
			mock<ILogger>(),
		);

		expect(configUtil.load()).rejects.toThrowErrorMatchingSnapshot();
	});

	it("should reject paths outside of SPT root", () => {
		vol.fromNestedJSON(
			{
				"config.jsonc": `{
					"syncPaths": [
						"../../etc/shadow"
					],
					// Exclusions for commonly used SPT mods
					"exclusions": [
						"plugins/**/node_modules"
					]
				}`,
			},
			process.cwd(),
		);

		const configUtil = new ConfigUtil(
			new VFS() as IVFS,
			new JsonUtil() as IJsonUtil,
			new PreSptModLoader() as IPreSptModLoader,
			mock<ILogger>(),
		);

		expect(configUtil.load()).rejects.toThrowErrorMatchingSnapshot();
	});

	it("should reject non-array syncPaths", () => {
		vol.fromNestedJSON(
			{
				"config.jsonc": `{
					"syncPaths": "plugins",
					// Exclusions for commonly used SPT mods
					"exclusions": [
						"plugins/**/node_modules"
					]
				}`,
			},
			process.cwd(),
		);

		const configUtil = new ConfigUtil(
			new VFS() as IVFS,
			new JsonUtil() as IJsonUtil,
			new PreSptModLoader() as IPreSptModLoader,
			mock<ILogger>(),
		);

		expect(configUtil.load()).rejects.toThrowErrorMatchingSnapshot();
	});

	it("should warn on explicitly excluded syncPaths", () => {
		vol.fromNestedJSON(
			{
				"config.jsonc": `{
					"syncPaths": [
						"plugins"
					],
					// Exclusions for commonly used SPT mods
					"exclusions": [
						"plugins"
					]
				}`,
			},
			process.cwd(),
		);

		const logger = {
			warning: vi.fn((...args) => { }),
		} as unknown as ILogger;

		const configUtil = new ConfigUtil(
			new VFS() as IVFS,
			new JsonUtil() as IJsonUtil,
			new PreSptModLoader() as IPreSptModLoader,
			logger,
		);

		expect(configUtil.load()).rejects.toThrowErrorMatchingSnapshot()
	})

	it("should reject on non-array exclusions", () => {
		vol.fromNestedJSON(
			{
				"config.jsonc": `{
					"syncPaths": [
						"plugins"
					],
					// Exclusions for commonly used SPT mods
					"exclusions": "plugins/**/node_modules"
				}`,
			},
			process.cwd(),
		);

		const configUtil = new ConfigUtil(
			new VFS() as IVFS,
			new JsonUtil() as IJsonUtil,
			new PreSptModLoader() as IPreSptModLoader,
			mock<ILogger>(),
		);

		expect(configUtil.load()).rejects.toThrowErrorMatchingSnapshot();
	});

	it("should reject invalid JSON", () => {
		vol.fromNestedJSON(
			{
				"config.jsonc": `{
					"invalid": "invalid"
				}`,
			},
			process.cwd(),
		);

		const configUtil = new ConfigUtil(
			new VFS() as IVFS,
			new JsonUtil() as IJsonUtil,
			new PreSptModLoader() as IPreSptModLoader,
			mock<ILogger>(),
		);

		expect(configUtil.load()).rejects.toThrowErrorMatchingSnapshot();
	});
});
