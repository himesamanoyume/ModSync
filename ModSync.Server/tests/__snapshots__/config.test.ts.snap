// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ConfigUtil > Should always include ModSync.Updater.exe in syncPaths 1`] = `
[
  {
    "enabled": true,
    "enforced": true,
    "path": "ModSync.Updater.exe",
    "restartRequired": false,
    "silent": true,
  },
  {
    "enabled": true,
    "enforced": true,
    "path": "BepInEx/plugins/Corter-ModSync.dll",
    "restartRequired": true,
    "silent": true,
  },
]
`;

exports[`ConfigUtil > should load config 1`] = `
[
  {
    "enabled": true,
    "enforced": true,
    "path": "ModSync.Updater.exe",
    "restartRequired": false,
    "silent": true,
  },
  {
    "enabled": true,
    "enforced": true,
    "path": "BepInEx/plugins/Corter-ModSync.dll",
    "restartRequired": true,
    "silent": true,
  },
  {
    "enabled": true,
    "enforced": false,
    "path": "doesnotexist",
    "restartRequired": true,
    "silent": false,
  },
  {
    "enabled": true,
    "enforced": false,
    "path": "plugins",
    "restartRequired": true,
    "silent": false,
  },
  {
    "enabled": false,
    "enforced": false,
    "path": "mods",
    "restartRequired": true,
    "silent": false,
  },
]
`;

exports[`ConfigUtil > should reject absolute paths 1`] = `[Error: Corter-ModSync: SyncPaths must be relative to SPT server root. Invalid path '/etc/shadow']`;

exports[`ConfigUtil > should reject invalid JSON 1`] = `[Error: Corter-ModSync: config.jsonc 'syncPaths' is not an array. Please verify your config is correct and try again.]`;

exports[`ConfigUtil > should reject non-array syncPaths 1`] = `[Error: Corter-ModSync: config.jsonc 'syncPaths' is not an array. Please verify your config is correct and try again.]`;

exports[`ConfigUtil > should reject on non-array exclusions 1`] = `[Error: Corter-ModSync: config.jsonc 'exclusions' is not an array. Please verify your config is correct and try again.]`;

exports[`ConfigUtil > should reject paths outside of SPT root 1`] = `[Error: Corter-ModSync: SyncPaths must within SPT server root. Invalid path '../../etc/shadow']`;

exports[`ConfigUtil > should reject syncPath object without path 1`] = `[Error: Corter-ModSync: config.jsonc 'syncPaths' is missing 'path'. Please verify your config is correct and try again.]`;

exports[`ConfigUtil > should warn on explicitly excluded syncPaths 1`] = `[Error: Corter-ModSync: 'plugins' has been added as a sync path and is also in the 'exclusions' array. This probably isn't doing what you want. If you no longer want to sync this path, remove it from the 'exclusions' and 'syncPaths' arrays.]`;
